src/InvoiceManager.php

```diff
@@ -0,0 +1,326 @@
+            }
+            if (InvoiceStatus::PAID == $second_invoice->getStatus()) {
+                throw new InvalidInvoiceStatusException();
+            }
+            if ($first_invoice->getUserId() != $second_invoice->getUserId()) {
+                throw new InvoiceUserMismatchException();
+            }
+            if ($first_invoice->getCurrencyId() !== $second_invoice->getCurrencyId()) {
+                throw new CurrencyMismatchException();
+            }
+
+            return $this->createInvoice($first_invoice, $second_invoice, $localizedDetails);
+        });
+    }
+
+    protected function createInvoice(Invoice $first_invoice, Invoice $second_invoice, array $localizedDetails): IInvoice
+    {
+        $invoice = new Invoice();
+        $invoice->user_id = $first_invoice->getUserId();
+        $invoice->currency_id = $first_invoice->getCurrencyId();
+        if (null != $first_invoice->getMeta() || null != $second_invoice->getMeta()) {
+            $invoice->meta = array_merge($first_invoice->getMeta() ?? [], $second_invoice->getMeta() ?? []);
+        }
+        $invoice->amount = $first_invoice->amount->getValue() + $second_invoice->amount->getValue();
+        if (isset($localizedDetails['title'])) {
+            $invoice->title = $localizedDetails['title'];
+        }
+        $invoice->save();
+        $first_invoice->products()
+                      ->update([
+                                   'invoice_id' => $invoice->getID(),
+                               ]);
+        $second_invoice->products()
+                       ->update([
+                                    'invoice_id' => $invoice->getID(),
+                                ]);
+
+        return $invoice;
+    }
+
+    public function addPaymentToInvoice(int $invoiceId, string $type, INumber $amount, PaymentStatus $status, ?array $meta): IPayment
+    {
+        // TODO: Implement addPaymentToInvoice() method.
+        $invoice = $this->getInvoiceById($invoiceId);
+        if (InvoiceStatus::PAID == $invoice->getStatus()) {
+            throw new InvalidInvoiceStatusException();
+        }
+        $payment = $invoice->payments()
+                           ->create([
+                                        'method' => $type,
+                                        'amount' => $amount,
+                                        'meta' => $meta,
+                                        'status' => $status,
+                                    ]);
+
+        return $payment;
+    }
+
+    public function approvePayment(int $paymentId): IPayment
+    {
+        // TODO: Implement approvePayment() method.
+        $payment = Payment::query()
+                          ->findOrFail($paymentId);
+        if (InvoiceStatus::PAID == $payment->invoice->getStatus()) {
+            throw new InvalidInvoiceStatusException();
+        }
+        if ($payment->getAmount() > $payment->invoice->getAmount()) {
+            throw new AmountInvoiceMismatchException();
```

Comment:
You check amount of payment after you inserted in the database?

Reply:
done